umi_tools/whitelist.py

```diff
@@ -0,0 +1,456 @@
+
+        if not options.pattern2:
+            options.pattern2 = options.pattern
+
+    extract_cell = False
+    extract_umi = False
+
+    # If the pattern is a regex we can compile the regex(es) prior to
+    # ExtractFilterAndUpdate instantiation
+    if options.extract_method == "regex":
+        if options.pattern:
+            try:
+                options.pattern = regex.compile(options.pattern)
+            except regex.error:
+                U.error("barcode_regex '%s' is not a "
+                        "valid regex" % options.pattern)
+
+        if options.pattern2:
+            try:
+                options.pattern2 = regex.compile(options.barcode_regex2)
+            except regex.Error:
+                U.error("barcode_regex2 '%s' is not a "
+                        "valid regex" % options.barcode_regex2)
+
+    # check whether the regex contains a umi group(s) and cell groups(s)
+    if options.extract_method == "regex":
+        if options.pattern:
+            for group in options.pattern.groupindex:
+                if group.startswith("cell_"):
+                    extract_cell = True
+                elif group.startswith("umi_"):
+                    extract_umi = True
+        if options.pattern2:
+            for group in options.pattern2.groupindex:
+                if group.startswith("cell_"):
+                    extract_cell = True
+                elif group.startswith("umi_"):
+                    extract_umi = True
+
+    # check whether the pattern string contains umi/cell bases
+    elif options.extract_method == "string":
+        if options.pattern:
+            if "C" in options.pattern:
+                extract_cell = True
+            if "N" in options.pattern:
+                extract_umi = True
+        if options.pattern2:
+            if "C" in options.pattern2:
+                extract_cell = True
+            if "N" in options.pattern2:
+                extract_umi = True
+
+    if not extract_umi:
+        if options.extract_method == "string":
+            U.error("barcode pattern(s) do not include any umi bases "
+                    "(marked with 'Ns') %s, %s" % (
+                        options.pattern, options.pattern2))
+        elif options.extract_method == "regex":
+            U.error("barcode regex(es) do not include any umi groups "
+                    "(starting with 'umi_') %s, %s" (
+                        options.pattern, options.pattern2))
+    if not extract_cell:
+        if options.extract_method == "string":
+            U.error("barcode pattern(s) do not include any cell bases "
+                    "(marked with 'Cs') %s, %s" % (
+                        options.pattern, options.pattern2))
+        elif options.extract_method == "regex":
+            U.error("barcode regex(es) do not include any cell groups "
+                    "(starting with 'cell_') %s, %s" (
+                        options.pattern, options.pattern2))
+
+    if options.stdin == sys.stdin:
```

Comment:
You're right, this is uneccessary since `options.stdin` is handled by U.start such that it will always be a file object so lines 356-359 can be replaced by `umi_methods.fastqIterate(options.stdin)` . This was copied from umi_tools.extract so I'll remedy there too.